main-command/src/main/scala/sbt/internal/client/NetworkClient.scala

```diff
@@ -9,94 +9,242 @@ package sbt
 package internal
 package client
 
-import java.io.{ File, IOException }
+import java.io.{ File, IOException, InputStream, PrintStream }
+import java.lang.ProcessBuilder.Redirect
+import java.net.Socket
+import java.nio.channels.ClosedChannelException
+import java.nio.file.Files
 import java.util.UUID
 import java.util.concurrent.atomic.{ AtomicBoolean, AtomicReference }
+import java.util.concurrent.{ ConcurrentHashMap, LinkedBlockingQueue, TimeUnit }
 
+import sbt.internal.client.NetworkClient.Arguments
 import sbt.internal.langserver.{ LogMessageParams, MessageType, PublishDiagnosticsParams }
 import sbt.internal.protocol._
-import sbt.internal.util.{ ConsoleAppender, LineReader }
+import sbt.internal.util.{ ConsoleAppender, ConsoleOut, Signals, Terminal, Util }
 import sbt.io.IO
 import sbt.io.syntax._
 import sbt.protocol._
 import sbt.util.Level
+import sjsonnew.BasicJsonProtocol._
+import sjsonnew.shaded.scalajson.ast.unsafe.{ JObject, JValue }
 import sjsonnew.support.scalajson.unsafe.Converter
 
-import scala.collection.mutable.ListBuffer
-import scala.sys.process.{ BasicIO, Process, ProcessLogger }
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.concurrent.duration._
 import scala.util.control.NonFatal
-import scala.util.{ Failure, Success }
+import scala.util.{ Failure, Properties, Success }
+import Serialization.{
+  CancelAll,
+  attach,
+  cancelRequest,
+  promptChannel,
+  systemIn,
+  systemOut,
+  terminalCapabilities,
+  terminalCapabilitiesResponse,
+  terminalPropertiesQuery,
+  terminalPropertiesResponse
+}
+import NetworkClient.Arguments
+
+trait ConsoleInterface {
+  def appendLog(level: Level.Value, message: => String): Unit
+  def success(msg: String): Unit
+}
 
-class NetworkClient(configuration: xsbti.AppConfiguration, arguments: List[String]) { self =>
-  private val channelName = new AtomicReference("_")
+class NetworkClient(
```

Comment:
Is this NetworkClient being double dipped? Once for the thin client v1 (https://github.com/sbt/sbt/pull/2859) and now for this? 

For debugging, documentation etc, I think it would be helpful to somehow distinguish the GraalVM-built version on the welcome tag, like:

```
$ sbtc
[info] entering *experimental* native thin client - BEEP WHIRR
```

vs

```
$ sbt -client
[info] entering *experimental* JVM thin client - BEEP WHIRR
```

Is the intent to keep the old `-client` for non-GraalVM fallback, or build another JAR for the purpose?

Reply:
I tried to leave NetworkClient compatible with `sbt -client` simply because I could. I figured it would be nice if `sbt -client` would continue working. That being said, I wouldn't mind getting rid of it. For non-GraalVM fallback, I think we could pretty easily make a fat jar. We could also probably use the trick that mill uses to embed a bash or batch script in the fat jar to make it executable.

Reply:
Thin client v1 was shipped as a command, so it's in main-command subproject. If v2 lives as a separate JAR, we could pull it out as a subproject is what I was thinking.

Reply:
Got it. It would be pretty easy to move NetworkClient.scala. I also think we could move `Server` and `ServerConnection` to main which I think makes a bit more sense.